src/mpid/ch3/channels/nemesis/netmod/tcp/tcp_init.c

```diff
@@ -374,29 +385,75 @@ static int GetSockInterfaceAddr(int myRank, char *ifname, int maxIfname,
 
     /* If we don't have an IP address, try to get it from the name */
     if (!ifaddrFound) {
-        int i;
-	struct hostent *info = NULL;
+        int i, status = 0, ipv4_found = 0, ipv6_found = 0;
+        struct addrinfo hints;
+        struct addrinfo *result, *rp;
+        memset(&hints, 0, sizeof(struct addrinfo));
+        hints.ai_family = AF_UNSPEC;
+        hints.ai_socktype = SOCK_STREAM;
+        hints.ai_protocol = 0;
+        hints.ai_flags = 0;
+        /* First try getting IPV4 network address */
         for (i = 0; i < MPIR_CVAR_NEMESIS_TCP_HOST_LOOKUP_RETRIES; ++i) {
-            info = gethostbyname( ifname_string );
-            if (info || h_errno != TRY_AGAIN)
+            status = getaddrinfo(ifname_string, NULL, &hints, &result);
+            /* If connection failed, skip to next try */
+            if (status != 0)
+                continue;
+            /* Only grep those IPV4 address*/
+            for (rp = result; rp; rp = rp->ai_next) {
+                if (rp->ai_family == AF_INET) {
+                    ifaddr->len  = rp->ai_addrlen;
+                    ifaddr->type = AF_INET;
+                    if (ifaddr->len > sizeof(ifaddr->ifaddr)) {
+                        ifaddr->len = 0;
+                        ifaddr->type = -1;
+                        MPIR_ERR_INTERNAL(mpi_errno, "Address too long to fit in field");
+                    }
+                    else {
+                        MPIR_Memcpy( ifaddr->ifaddr, &(((struct sockaddr_in* )rp->ai_addr)->sin_addr), ifaddr->len);
+                        /* If we found the address and successfully copied it break from current address loop and retry loop*/
+                        ipv4_found = 1;
+                        break;
+                    }
+                }
+            }
+            if (ipv4_found)
                 break;
         }
-        MPIR_ERR_CHKANDJUMP2(!info || !info->h_addr_list, mpi_errno, MPI_ERR_OTHER, "**gethostbyname", "**gethostbyname %s %d", ifname_string, h_errno);
-        
-        /* Use the primary address */
-        ifaddr->len  = info->h_length;
-        ifaddr->type = info->h_addrtype;
-        if (ifaddr->len > sizeof(ifaddr->ifaddr)) {
-            /* If the address won't fit in the field, reset to
-               no address */
-            ifaddr->len = 0;
-            ifaddr->type = -1;
-            MPIR_ERR_INTERNAL(mpi_errno, "Address too long to fit in field");
-        } else {
-            MPIR_Memcpy( ifaddr->ifaddr, info->h_addr_list[0], ifaddr->len );
-	}
+        /* If we unable to call getaddrinfo, we break from retry loop*/
+        MPIR_ERR_CHKANDJUMP2(status != 0, mpi_errno, MPI_ERR_OTHER, "**getaddrinfo", "**getaddrinfo %s %d", ifname_string, h_errno);
```

Comment:
The `**getaddrinfo` and `**getaddrinfo %s %d` error names should be added to `src/mpi/errhan/errnames.txt`. They should replace `**gethostbyname` and `**gethostbyname %s %d` in that file.